### ISSUE TITLE
Janky first few frames of simple width/height/transformX/transformY animation

### ISSUE THREAD

**Issue by drewandre** [software-mansion/react-native-reanimated#1295]
First off, reanimated2 has been a game changer for my project. Thank you to Software Mansion for all their hard work as well as the awesome seminars they've been putting on!

## Description
I've been building out `<PopView />` component that behaves similarly to iOS control center modules. When activated, the module's width and height expand to a pre-defined maxWidth/maxHeight, and the module's view is translated so as to place it directly in the center of the screen. Users can tap outside the module to close using a full screen semi-opaque view behind the module. The animation's width, height, translateX and translateY values are all derived from a single `progress` animated value which I thought to be the best approach to keep everything proportional and smooth.

The calculations are correction, however the first few frames jump around a lot when using `withTiming` and it's default config. Using `withSpring` and the default spring config is significantly worse, and actually renders the animation almost unusable. It's almost like 60fps isn't enough for this animation, and running at higher speeds (like when using `withSpring` or a shorter `withTiming` duration) just results in skipped frames.

I guess what I'm really asking is: hypothetically, if I animate a view from the top of the screen to the bottom using `withTiming` and a crazy short duration of like 5ms, frames are going to be skipped, right? The view can't possibly be moved pixel-by-pixel on it's way to the bottom, otherwise give an animation duration of 5ms and a window height of 812 (iphone Xs max) you'd need an extremely high fps?

If it's hypothetically possible, my next questions are:
- Does my approach make sense (all four layout values derived from `progress`)?
- Should I consider running the animation progress from 0-100 instead of 0-1...?
- Could there be an underlying issue with float calculations in reanimated2?
- Should I consider implementing some sort of `requestAnimationFrame`?
- If LayoutAnimation could animate a view without this issue (I've tried), is there any reason why reanimated wouldn't be able to?

Below is a screen recording below where you can see 1) the full animation run, and then 2) a slowed down version showing the jumping around I'm seeing in the first few frames. I used `withSpring` to really highlight the issue.

![ezgif com-gif-maker](https://user-images.githubusercontent.com/13245041/94582376-05451a80-0231-11eb-92b9-b74a1e91900c.gif)

## Code
Relevant code is pasted below, but a full repo demonstrating the issue can be found here (and is just a fork of the 2.0.0-alpha.7 playground repo).

```js
export function Box({
  id,
  position: {
    maxWidth = windowWidth * 0.9,
    maxHeight = windowHeight * 0.5,
    height,
    width,
    x,
    y
  },
  progress,
  activeModuleId,
  children
}) {

  function calculateProgress(p) {
    'worklet';
    if (activeModuleId.value === id || activeModuleId.value === null) {
      return p
    } else {
      return 0
    }
  }

  function calculateOpacity(p) {
    'worklet';
    if (activeModuleId.value === id || activeModuleId.value === null) {
      return 1
    } else {
      return 1 - p
    }
  }

  function calculateZ() {
    'worklet';
    if (activeModuleId.value === id || activeModuleId.value === null) {
      return 9999
    } else {
      return 0
    }
  }

  const moduleWidthOffset = maxWidth - width
  const moduleWidth = useDerivedValue(() => width + calculateProgress(moduleWidthOffset * progress.value))

  const moduleHeightOffset = maxHeight - height
  const moduleHeight = useDerivedValue(() => height + calculateProgress(moduleHeightOffset * progress.value))

  const translateXOffset = -x + (windowWidth - maxWidth) * 0.5
  const translateX = useDerivedValue(() => calculateProgress(translateXOffset * progress.value))

  const translateYOffset = -y + (windowHeight - maxHeight) * 0.5
  const translateY = useDerivedValue(() => calculateProgress(translateYOffset * progress.value))

  const animatedBoxStyles = useAnimatedStyle(() => ({
    top: y,
    left: x,
    width: moduleWidth.value,
    height: moduleHeight.value,
    zIndex: calculateZ(progress.value),
    opacity: calculateOpacity(progress.value),
    transform: [
      { translateX: translateX.value },
      { translateY: translateY.value },
    ]
  }))

  const onModuleTapGestureHandler = useAnimatedGestureHandler({
    onStart: () => {
      activeModuleId.value = id
      progress.value = withTiming(1, withTimingConfig)
    },
  })

  return (
    <TapGestureHandler onGestureEvent={onModuleTapGestureHandler}>
      <Animated.View style={[styles.box, animatedBoxStyles]}>
        {children}
      </Animated.View>
    </TapGestureHandler>
  );
}
```

### Package versions
- React: 16.13.1
- React Native: 0.63.1
- React Native Reanimated: 2.0.0-alpha.7

**Comment by terrysahaidak** [comment]
Could you please test your example but with changing [this line](https://github.com/software-mansion/react-native-reanimated/blob/master/src/reanimated2/animations.js#L216) to be

```js
animation.lastTimestamp = now - 16.66;
```

**Comment by drewandre** [comment]
@terrysahaidak I made that change but it didn't seem to make any difference. I attached a gif below. What was the intent behind that change?

![new](https://user-images.githubusercontent.com/13245041/94813495-ec617400-03ac-11eb-8390-222427ffe02b.gif)

**Comment by terrysahaidak** [comment]
Ok, just another issue I fixed with that change but seems to be not realted.

Why do you animate both y/x and translate? I think if you rewrite it to use only either translte of position it will fix it.

**Comment by drewandre** [comment]
Got it, worth a shot though. It looks like animating the x/y values instead of translating the entire view is working much better - buttery smooth. I'm going to mark this as closed unless you think this highlights some underlying issue with the translate transforms.

Thanks @terrysahaidak!

**Comment by terrysahaidak** [comment]
The reason behind this is that all layout props are animated asynchronously. But UI props – synchronously. So if your view depends on width/height – then it's better to animate top/left instead of translateY/translateX, so all the changes will happen in the same frame.

**Comment by terrysahaidak** [comment]
Hey @Szymon20000, is #1215 only about events?

**Comment by drewandre** [comment]
Thanks for the explanation @terrysahaidak -- I understand [layout props](https://reactnative.dev/docs/layout-props), but what do you mean by UI props?

**Comment by terrysahaidak** [comment]
UI props are props that can be updated on UI thread synchronously.
Basically, it's opacity, backgroundColor, and all the transforms.
All the properties currently available with useNativeDriver in Animated.

Here is the full list:
https://github.com/software-mansion/react-native-reanimated/blob/master/src/ConfigHelper.js#L6-L26

**Comment by drewandre** [comment]
Got it - so reanimated best practices would be to avoid mixing any of these properties in a `useAnimatedStyle` worklet?

**Comment by terrysahaidak** [comment]
Your case is more like an exception when you should not mix them because it's noticeable that they are async - because you transform the view and change its dimensions at the same time.

But, for example, changing them from an event handler will be sync because of #1215.

**Comment by drewandre** [comment]
Understood, thanks again for your explanation. Closing this issue now and have a nice weekend!

**Comment by Szymon20000** [comment]
> Hey @Szymon20000, is #1215 only about events?

No, it has an impact on every non-UI prop update. However, there are still cases when we cannot update those props synchronously.
